File: cpp/47. Permutations II.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
/*
Ex: [1, 1, 2]

    1     x     2
   . .         . .
  1   2       1   x
  .   .       .
  2   1       1

A key insight to avoid generating any redundant permutation is that at each step rather than viewing each number as a candidate, 
we consider each unique number as the true candidate. 
For instance, at the very beginning, given in the input of [1, 1, 2], we have only two true candidates instead of three.
*/
public:
    vvi ans;
    void dfs(int i, vi &path, unordered_set<int> &done, vi &nums) {
        int n = nums.size();
        if(i==n) {
            ans.push_back(path);
            return;
        }
        unordered_set<int> checkDup;
        for(int j = 0; j < n; j++) {
            if(done.count(j) || checkDup.count(nums[j])) continue;
            done.insert(j);
            checkDup.insert(nums[j]);
            path.push_back(nums[j]);
            dfs(i+1, path, done, nums);
            done.erase(j);
            path.pop_back();
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vi path;
        unordered_set<int> done;
        dfs(0, path, done, nums);
        return ans;
    }
};

void printAns(vvi ans) {
    for(auto l: ans) {
        cout << "(";
        for(auto n: l) {
            cout << n << " ";
        }
        cout << "), ";
    }
    cout << "\n";
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; vvi ans;

    solution = Solution();
    nums = {1,1,2};
    ans = solution.permuteUnique(nums);
    printAns(ans);

    solution = Solution();
    nums = {1,2,3};
    ans = solution.permuteUnique(nums);
    printAns(ans);
}

File: python/47-Permutations II.py
from typing import List

'''
Ex: [1, 1, 2]

    1     x     2
   . .         . .
  1   2       1   x
  .   .       .
  2   1       1

A key insight to avoid generating any redundant permutation is that at each step rather than viewing each number as a candidate, 
we consider each unique number as the true candidate. 
For instance, at the very beginning, given in the input of [1, 1, 2], we have only two true candidates instead of three.
'''

class Solution:
    # def permuteUnique(self, nums: List[int]):
    #     "Time: O(n!), Space: O(n)"
    #     n = len(nums)
    #     result = set()
    #     done = set()
    #     def helper(path):
    #         if len(path) == n:
    #             result.add(path)
    #             return
    #         for i, num in enumerate(nums):
    #             if i not in done:
    #                 done.add(i)
    #                 helper(path+(num,))
    #                 done.remove(i)
    #     helper(())
    #     return result
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        "Improvement of above, Time: O(n!), Space: O(n)"
        n = len(nums)
        result = []
        done = set()
        def helper(path):
            if len(path) == n:
                result.append(path)
                return
            check_dup = set()
            for i, num in enumerate(nums):
                if i not in done and num not in check_dup:
                    done.add(i)
                    check_dup.add(num)
                    helper(path+[num])
                    done.remove(i)
        helper([])
        return result

nums = [1,1,2]
result = Solution().permuteUnique(nums)
print(result) # [[1,1,2], [1,2,1], [2,1,1]]

nums = [1,2,3]
result = Solution().permuteUnique(nums)
print(result) # [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]